### PR TITLE
Make it work, add ability to mark current state

### PR DESCRIPTION
**Pull request by haffla**
Hello, I believe this actually didn't work anymore. I kept getting `ArgumentError` for `Stateoscope::Adapter::Base#initialize`.  Is it still maintained? 

I made it work again and added some sugar. You can now configure the output format via `options` and also pass a `current_state` which node will be coloured green. I believe this can be very helpful.

### CODE REVIEW

**Comment by padde** [comment]
Hey @haffla thanks for your contribution! I must have overlooked your PR, sorry it took so long. I will take a look soon and release a new version. First I want to fix the CI build and update some of the development dependencies though.

**Comment by haffla** [comment]
cool I'll update the code to tame the houndci-bot

**Comment by padde** [comment]
@haffla build is working again, please merge/rebase current `master`.

**Comment by haffla** [comment]
I've fixed the failing tests, not sure why I had overlooked the test suite.

They mostly failed because now that `Base` is not a struct anymore, but a class, it's constructor parameters are actually required. The overall api of the gem did not change.

Also would it make sense to add Rubocop and enforce a style + add check to Travis?

I will update the Readme, too.

File: spec/lib/stateoscope/tasks/stateoscope_rake_spec.rb [hunk 1]
```diff
@@ -41,5 +41,14 @@
...
     end
+
+    context 'with three arguments' do
+      it 'invokes Stateoscope.visualize with the supplied class and state_machine_name and current_state' do
+        stateoscope = class_double('Stateoscope').as_stubbed_const
+        allow(stateoscope).to receive(:visualize)
+        rake['stateoscope:visualize'].invoke('DummyClass', 'foo', 'bar')
+        expect(stateoscope).to have_received(:visualize).with(DummyClass, state_machine_name: 'foo', current_state: 'bar')
+      end
+    end
```
Comment (padde) [inline spec/lib/stateoscope/tasks/stateoscope_rake_spec.rb:52]: It would be helpful if you could add this variant to the README for documentation purposes.

File: lib/tasks/stateoscope.rake [hunk 2]
```diff
@@ -1,8 +1,10 @@
...
   desc 'visualize state machine for a given class'
-  task :visualize, [:class, :state_machine_name] => :environment do |t, args|
-    fail ArgumentError, 'missing required argument <class>' unless args[:class].present?
+  task :visualize, %i[class state_machine_name current_state] => :environment do |_, args|
+    raise ArgumentError, 'missing required argument <class>' if args[:class].nil?
     klass = args[:class].classify.constantize
-    Stateoscope.visualize(klass, state_machine_name: args[:state_machine_name])
+    # class not needed in args
+    args = args.reject { |k, _| k == :class }.to_h
+    Stateoscope.visualize(klass, args)
```
Comment (padde) [inline lib/tasks/stateoscope.rake]: I believe this can be improved by using `Hash#delete`. It is more concise and we don't need to repeat the`:class` key:

    klass = args.delete(:class).classify.constantize
    Stateoscope.visualize(klass, args)

File: lib/tasks/stateoscope.rake [hunk 3]
```diff
@@ -1,8 +1,10 @@
 namespace :stateoscope do
   desc 'visualize state machine for a given class'
-  task :visualize, [:class, :state_machine_name] => :environment do |t, args|
-    fail ArgumentError, 'missing required argument <class>' unless args[:class].present?
+  task :visualize, %i[class state_machine_name current_state] => :environment do |_, args|
+    raise ArgumentError, 'missing required argument <class>' if args[:class].nil?
```
Comment (padde) [inline lib/tasks/stateoscope.rake]: `if x.nil?` is not exactly the same as `unless x.present?` - it now allows blank strings. I would prefer that a blank string still trigger the exception here. I think `if x.blank?` will do the trick.

File: lib/stateoscope/visualizer.rb [hunk 4]
```diff
@@ -10,8 +10,11 @@ def parse_graph
       add_state_transitions
     end
 
-    def output(filename)
-      @viz.output(pdf: "#{filename}.pdf")
+    def output(filename, format)
+      options = {}
+      options[format] = "#{filename}.#{format}"
+      @viz.output(options)
+      "#{filename}.#{format}"
```
Comment (padde) [inline lib/stateoscope/visualizer.rb]: Why is this line necessary?

File: lib/stateoscope/visualizer.rb [hunk 5]
```diff
@@ -10,8 +10,11 @@ def parse_graph
       add_state_transitions
     end
 
-    def output(filename)
-      @viz.output(pdf: "#{filename}.pdf")
+    def output(filename, format)
+      options = {}
+      options[format] = "#{filename}.#{format}"
+      @viz.output(options)
```
Comment (padde) [inline lib/stateoscope/visualizer.rb]: This can be boiled down to

```
@viz.output(format => "#{filename}.#{format}")
```

However, I don't like how the responsibility of generating the filename is now split across two classes. I'd rather add a `format` argument to `Stateoscope.filename_for` so we can simply use the full filename here. What do you think?

BTW I realize it was that way before, my bad 😉

**Review by padde — CHANGES_REQUESTED** [review]
Thanks!

> Also would it make sense to add Rubocop and enforce a style + add check to Travis?

I would see that as a separate issue. Personally, I think hound is enough but just skimping over the warnings in this PR I must say that I don't agree with a lot of the suggestions and might disable it completely.

> I will update the Readme, too.

That would  be awesome, esp. for the case I mentioned in the review comments.

Other than that I have just some style comments.

File: lib/tasks/stateoscope.rake (hunk 2 above)
Comment (haffla) [reply lib/tasks/stateoscope.rake]: yea I tried `Hash#delete` but didn't work. I actually just found out that it's not a `Hash` but a `Rake::TaskArguments`. I think it would actually be best to have `Stateoscope.visualize` just take one argument (a `Hash`), but I'm not gonna change that now.

File: lib/tasks/stateoscope.rake (hunk 3 above)
Comment (haffla) [reply lib/tasks/stateoscope.rake]: Yes sorry not sure why I changed that. I'm working with vscode and a rubocop extension and probably it complained about that part because `unless something` can always be rewritten with `if the opposite`. Sure `nil?` is not the opposite of `present?` I know. 

By the way `blank?`, `present?`, `classify`, `constantize` are all Rails. Have you tried to use the gem outside of Rails?

File: lib/stateoscope/visualizer.rb (hunk 4 above)
Comment (haffla) [reply lib/stateoscope/visualizer.rb]: I'm using Stateoscope not through rake but dynamically in a Rails project. In one of my controllers I do

```
filepath = Stateoscope.visualize(
        BaustarterState,
        current_state: @user.baustarter_state.state,
        dir: Rails.root.join("public"),
        format: "png"
      )

# return filepath to client
```

Anyway since the filename is created from a timestamp in `Stateoscope.filename_for` it would be nice to return the final filename to the caller, otherwise they have no idea where the just created file is located.

File: lib/stateoscope/visualizer.rb (hunk 5 above)
Comment (haffla) [reply lib/stateoscope/visualizer.rb]: you are right. gonna change it

File: lib/stateoscope/visualizer.rb (hunk 4 above)
Comment (haffla) [reply lib/stateoscope/visualizer.rb]: ok yea I removed it from the visualizer and just return the filename from `visualize` method. Better.

**Comment by haffla** [comment]
@padde thanks for the feedback. Right, there were some things that could be improved. I refactored the code a bit, added more documentation and specs.

File: lib/tasks/stateoscope.rake (hunk 3 above)
Comment (padde) [reply lib/tasks/stateoscope.rake]: Yes I thought about that as well while writing this comment but at the moment there is a dependency on activesupport in the gemfile anyway. We might want to get rid of that in the future, see #7

**Review by padde — APPROVED** [review]
Awesome work, thank you 👌

I think the whole file name logic could generally use a small overhaul but I‘ll tackle that as a separate issue. If you’re interested I can mention you in the PR.

**Comment by haffla** [comment]
Great! Will follow the progress! :)

**Comment by padde** [comment]
@haffla I've *finally* released v0.1.2 😄
